### PR TITLE
feat: Exclude Objects

### LINKED ISSUE #359: Exclude Region
## Thank you for using fluidd!

Nono! Thank YOU  for all your hard work!

## Is your feature request related to a problem? Please describe.
The main plugin I miss from my Octoprint/Klipper days is the "exclude region" one.

## Describe the solution you'd like:
I'd like to see it implemented as part of the GCode Preview box,.

## Describe alternatives you've considered:
There are no alternatives that I know of.

## Affected users and/or setups:
Users that have ever had a print consisting of many objects, and had one get messed up, will love you for it.

### PR DESCRIPTION
**Pull request by matmen**
Closes #359 
Credits and big thanks to @troy-jacobson and @kageurufu for doing most of the work here!

Using the Gcode viewer to cancel objects:
image
image
image
image

Alternatively using the cancel button at the top of the Gcode viewer:
image

### CODE REVIEW

**Comment by kageurufu** [comment]
Woohoo!

```
Signed-off-by: Troy Jacobson <troy.d.jacobson@gmail.com>
Co-authored-by: Franklyn Tackitt <git@frank.af>
```

File: src/components/widgets/gcode-preview/GcodePreviewCard.vue [hunk 1]
```diff
@@ -17,6 +17,18 @@
         </app-btn>
       </app-btn-collapse-group>
 
+      <app-btn
+        color=""
+        class="ml-1"
+        fab
+        x-small
+        text
+        :disabled="!parts.length"
```
Comment (pedrolamas) [inline src/components/widgets/gcode-preview/GcodePreviewCard.vue:26]: This didn't work for me until I changed it to this:
```suggestion
        :disabled="parts.length === 0"
```

File: src/store/parts/mutations.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    if ('excluded_objects' in payload) {
+      Vue.set(state, 'excludedParts', payload.excluded_objects)
+    }
+
+    if ('objects' in payload) {
+      const partMap: { [key: string]: Part} = {}
+      payload.objects.forEach((obj: PartObject) => {
+        const name = obj.name
+        const part: Part = {
+          name: name,
```
Comment (pedrolamas) [inline src/store/parts/mutations.ts:28]:
```suggestion
          name,
```

File: src/components/widgets/gcode-preview/GcodePreviewCard.vue (hunk 1 above)
Comment (pedrolamas) [reply src/components/widgets/gcode-preview/GcodePreviewCard.vue:26]: Actually, this is incorrect too, as parts is an object, not an array!

Comment (pedrolamas) [inline src/components/widgets/gcode-preview/GcodePreviewCard.vue:26]: Start a print, wait for this to enable, press CANCEL... still enabled!
image
```suggestion
        :disabled="!klippyReady || !printerPrinting || !printerFileLoaded || !parts.length"
```

**Review by pedrolamas — CHANGES_REQUESTED** [review]
Something feels off about this dialog! 
image

**Comment by matmen** [comment]
> Something feels off about this dialog!

I agree, it doesn't look very great, but I'm not sure how else to style it. If you have any suggestions please let me know ;)
Maybe dynamic width of the dialog as well as having a fixed size for the actions column would be enough here, I'll need to try that

File: src/components/widgets/gcode-preview/GcodePreview.vue [hunk 3]
```diff
@@ -276,6 +285,19 @@ export default class GcodePreview extends Mixins(StateMixin) {
     return this.panning ? 'optimizeSpeed' : 'geometricPrecision'
   }
 
+  get showExcludeObjects () {
+    const file = this.$store.getters['gcodePreview/getFile']
+    if (!file) {
+      return true
```
Comment (pedrolamas) [inline src/components/widgets/gcode-preview/GcodePreview.vue:291]: I believe this should be `false` (no file loaded, so don't show any exclude objects)
```suggestion
      return false
```
If true, I see the cancel objects even if the current file loaded is not the one printing!

**Comment by pedrolamas** [comment]
> If you have any suggestions please let me know

is the issue horizontal overflow? We could just set it to show "..." (and add a tooltip if you think it's worth it) in that case and revert the window size changes, no?

File: src/components/widgets/gcode-preview/GcodePreview.vue (hunk 3 above)
Comment (matmen) [reply src/components/widgets/gcode-preview/GcodePreview.vue]: Actually, this change doesn't seem right. It's there so you can cancel objects without having to have the gcode preview loaded. I tried this, and it looks like if the gcode preview has another file loaded it's disabled accordingly..

Comment (pedrolamas) [reply src/components/widgets/gcode-preview/GcodePreview.vue]: Ah I see!

However, it seems a bit weird to me to show the cancellation object on the gcode preview, but no gcode preview...

Comment (matmen) [reply src/components/widgets/gcode-preview/GcodePreview.vue]: I've tried it and UX wise it's fine I think.. And it has the big advantage of not having to parse the entire Gcode file on the client

File: src/components/widgets/exclude-objects/ExcludeObjects.vue [hunk 4]
```diff
@@ -41,30 +41,9 @@ import { Icons } from '@/globals'
...
-    const names = []
-    for (const name in parts) {
-      const excluded = this.isPartExcluded(name)
-      if (parts[name].target &&
-        ((state === 'printing' || state === 'paused') ||
-        ((state === 'complete' || state === 'cancelled' || state === 'error') && excluded))) {
-        names.push(name)
-      }
-    }
-    return names
```
Comment (matmen) [inline src/components/widgets/exclude-objects/ExcludeObjects.vue]: From my testing this isn't actually needed and causes some weird behavior where outlines are shown but icons aren't. Looking at the state the target and outline always seems to be set.

**Comment by matmen** [comment]
> is the issue horizontal overflow? We could just set it to show "..." (and add a tooltip if you think it's worth it) in that case and revert the window size changes, no?

Yep, that's exactly the issue. If you want to take a shot at it feel free to, my brain can't figure out how to use CSS properly right now 🤣

File: src/components/widgets/exclude-objects/ExcludeObjectsDialog.vue [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      </v-card-title>
+
+      <v-card-text class="py-3 px-5">
+        <v-simple-table>
+          <tbody>
+            <tr
+              v-for="part in parts"
+              :key="part.name"
+            >
+              <td class="partName">{{ part.name }}</td>
```
Comment (pedrolamas) [inline src/components/widgets/exclude-objects/ExcludeObjectsDialog.vue:24]: Linting:
```suggestion
              <td class="partName">
                {{ part.name }}
              </td>
```

**Review by pedrolamas — APPROVED** [review]

**Comment by pedrolamas** [comment]
Just tested the latest and feature wise everything seems to be working fine now! 🙂

**Pull request merged**